### DISCUSSION TITLE
why not a playwright plugin ?

### DISCUSSION THREAD

**Discussion by amk-dev** [antiwork/shortest#215] [Ideas]
### I was thinking, why build a seperate shortest api that have to be implement from scratch instead of building this as something people can plugin to already existing playwright instances

- tried a poc for the concept of shortest as a playwright plugin, it is pretty trivial to implement
- imo the value prop of `shortest` is not another `e2e` testing framework, its `ai powered QA`
- in that case extending playwright with ai capabilities can deliver more value to the user
- they can use all the battle tested testing goodness people usually expect from playwright
- and in the future `shortest` doesnt even have to be exclusive for playwright, packages like `shortest/vitest` / `shortest/jest` everything can exist ( they don't even have to be e2e )  
- modular, easy to plugin, and `shortest` dev team can focus on delivering value that did not exist previously ( all the ai stuff ), instead of reinventing a lot of wheels ( `hooks` / `showing test results` / `running failed tests` / `snapshots` etc etc )

in this demo, i write two tests, and use a playwright feature `test.only` that's very common in all the testing frameworks. just as a test. there will be so many features like this that people might expect. also all the community tools that comes with playwright, `shortest` can use right away.

https://github.com/user-attachments/assets/f272fb9b-469a-4468-90a5-95a3c14a782f

**Comment by slavingia** [comment]
Eventually shortest should replace playwright, that's the goal! But if this helps get mindshare, would be great

**Reply by amk-dev** [reply]
> Eventually shortest should replace playwright

💯

> But if this helps get mindshare, would be great

- absolutely. imo a plugin might be more likely to enter production codebases initially

**logistics**

- we can have a `@antiwork/shortest-plugin-playwright` package that is just this plugin
- or with the current package, we can have a path that imports the plugin, `@antiwork/shortest/plugins/playwright`, quiet long but it's a start

Note:

this is the POC i showed in the demo.

https://github.com/amk-dev/shortest-playwright-plugin-poc

**Reply by amk-dev** [reply]
I'll polish the demo, and show a usable demo, does that work ?

**Reply by slavingia** [reply]
Solid! Could we just do "shortest" without the .ai?

**Reply by amk-dev** [reply]
if there's any way, Will do it. Let me see

**Comment by amk-dev** [comment]
hey guys,

sorry that there was not much progress here. a lot of unexpected personal things happened ( my aunt passed away, my bag with my laptop and all accessories got stolen ). getting back to this this week.

while working on another project today, found out something that might be relevent here. so keeping a note of it.

**usecase**

1. I was writing a test to validate the creation of a gist using github api. 
2. But the response has a lot of dynamic values, like gist id, version etc
3. I wanted an integration test, so can't mock
4. Handling dynamic values where very challenging, i needed to know the shape of the response object, and the properties that are dynamic and write custom code to handle them.
5. So i added a new assertion to `expect`, which might also be possible to add to the expect we use in shortest
6. Assertions like this are very practical, since it doesnt cost much and definitely saves time and keeps tests super tidy


![vitest-expect-extend](https://github.com/user-attachments/assets/912263fe-5e21-4d0d-b4d5-9968550044d8)


![ray-so-export](https://github.com/user-attachments/assets/986ef074-6f1e-42e0-a1cf-afa7f9bb6b6b)



TLDR: Non e2e tests that need assertions on data might also be a good usecase for ai testing
